### ISSUE TITLE
armnn::IRuntime->EnqueueWorkload() causes app crash

### ISSUE THREAD

**Issue by the-promised-LAN** [ARM-software/armnn#456]
Hello!

**TLDR:**
armnn::IRuntime->EnqueueWorkload() crashes the Android Unity app without any explicit error messages 

**Details:**
Following one of the [ArmNN tutorials](https://community.arm.com/developer/tools-software/graphics/b/blog/posts/creating-ar-ml-demo-on-mobile-devices), I'm trying to get a trained [YOLOv1 Caffe model](https://drive.google.com/file/d/0Bx7QZuu7oVBbcWRpVG9NNl9EanM/view) working. The app crashes when calling `armnn::Status ret = s_Runtime->EnqueueWorkload(s_NetworkIdentifier, inTensor, outTensor);`, without any specific crash messaged in the adb logs. Call stack seems to point to `libc++_shared.__cxa_bad_cast` as the problem, though not sure.

Back trace:
```
09-14 15:26:30.633 24359 24389 E AndroidRuntime: 	at libc.abort(abort:160)
09-14 15:26:30.633 24359 24389 E AndroidRuntime: 	at libc++_shared.__cxa_bad_cast(__cxa_bad_cast:0)
09-14 15:26:30.633 24359 24389 E AndroidRuntime: 	at libc++_shared.0xa17e8(Native Method)
09-14 15:26:30.633 24359 24389 E AndroidRuntime: 	at libc++_shared.0xc8108(Native Method)
09-14 15:26:30.633 24359 24389 E AndroidRuntime: 	at libc++_shared.__cxa_get_exception_ptr(__cxa_get_exception_ptr:0)
09-14 15:26:30.633 24359 24389 E AndroidRuntime: 	at libc++_shared.0xc77ac(Native Method)
09-14 15:26:30.633 24359 24389 E AndroidRuntime: 	at libarmnn.std::__ndk1::unordered_map<int, std::__ndk1::unique_ptr<armnn::LoadedNetwork, std::__ndk1::default_delete<armnn::LoadedNetwork> >, std::__ndk1::hash<int>, std::__ndk1::equal_to<int>, std::__ndk1::allocator<std::__ndk1::pair<int const, std::__ndk1::unique_ptr<armnn::LoadedNetwork, std::__ndk1::default_delete<armnn::LoadedNetwork> > > > >::at(int const&) const(at:376)
09-14 15:26:30.633 24359 24389 E AndroidRuntime: 	at libarmnn.armnn::Runtime::GetLoadedNetworkPtr(int) const(GetLoadedNetworkPtr:80)
09-14 15:26:30.633 24359 24389 E AndroidRuntime: 	at libarmnn.armnn::Runtime::EnqueueWorkload(int, std::__ndk1::vector<std::__ndk1::pair<int, armnn::ConstTensor>, std::__ndk1::allocator<std::__ndk1::pair<int, armnn::ConstTensor> > > const&, std::__ndk1::vector<std::__ndk1::pair<int, armnn::Tensor>, std::__ndk1::allocator<std::__ndk1::pair<int, armnn::Tensor> > > const&)(EnqueueWorkload:48)
09-14 15:26:30.633 24359 24389 E AndroidRuntime: 	at libyoloDetector.detectObjects(detectObjects:200)
```

PC backtrace:
```
09-14 15:26:30.632 24359 24389 E AndroidRuntime: backtrace:
09-14 15:26:30.632 24359 24389 E AndroidRuntime:       #00 pc 00000000000832a8  /apex/com.android.runtime/lib64/bionic/libc.so (abort+160) (BuildId: b0750023d0cf44584c064da02400c159)
09-14 15:26:30.632 24359 24389 E AndroidRuntime:       #01 pc 00000000000a15ec  /data/app/com.DefaultCompany.ArmNNObject-ZZiJpDLPTj7kfZWZo6ik8Q==/lib/arm64/libc++_shared.so (__cxa_bad_cast) (BuildId: 95ba03fdc4a6fcf909919db0a725a514ed3fe2ff)
09-14 15:26:30.632 24359 24389 E AndroidRuntime:       #02 pc 00000000000a17e8  /data/app/com.DefaultCompany.ArmNNObject-ZZiJpDLPTj7kfZWZo6ik8Q==/lib/arm64/libc++_shared.so (BuildId: 95ba03fdc4a6fcf909919db0a725a514ed3fe2ff)
09-14 15:26:30.632 24359 24389 E AndroidRuntime:       #03 pc 00000000000c8108  /data/app/com.DefaultCompany.ArmNNObject-ZZiJpDLPTj7kfZWZo6ik8Q==/lib/arm64/libc++_shared.so (BuildId: 95ba03fdc4a6fcf909919db0a725a514ed3fe2ff)
09-14 15:26:30.632 24359 24389 E AndroidRuntime:       #04 pc 00000000000c782c  /data/app/com.DefaultCompany.ArmNNObject-ZZiJpDLPTj7kfZWZo6ik8Q==/lib/arm64/libc++_shared.so (__cxa_get_exception_ptr) (BuildId: 95ba03fdc4a6fcf909919db0a725a514ed3fe2ff)
09-14 15:26:30.632 24359 24389 E AndroidRuntime:       #05 pc 00000000000c77ac  /data/app/com.DefaultCompany.ArmNNObject-ZZiJpDLPTj7kfZWZo6ik8Q==/lib/arm64/libc++_shared.so (BuildId: 95ba03fdc4a6fcf909919db0a725a514ed3fe2ff)
09-14 15:26:30.632 24359 24389 E AndroidRuntime:       #06 pc 00000000006b8edc  /data/app/com.DefaultCompany.ArmNNObject-ZZiJpDLPTj7kfZWZo6ik8Q==/lib/arm64/libarmnn.so (std::__ndk1::unordered_map<int, std::__ndk1::unique_ptr<armnn::LoadedNetwork, std::__ndk1::default_delete<armnn::LoadedNetwork> >, std::__ndk1::hash<int>, std::__ndk1::equal_to<int>, std::__ndk1::allocator<std::__ndk1::pair<int const, std::__ndk1::unique_ptr<armnn::LoadedNetwork, std::__ndk1::default_delete<armnn::LoadedNetwork> > > > >::at(int const&) const+376)
09-14 15:26:30.632 24359 24389 E AndroidRuntime:       #07 pc 00000000006b4148  /data/app/com.DefaultCompany.ArmNNObject-ZZiJpDLPTj7kfZWZo6ik8Q==/lib/arm64/libarmnn.so (armnn::Runtime::GetLoadedNetworkPtr(int) const+80)
09-14 15:26:30.632 24359 24389 E AndroidRuntime:       #08 pc 00000000006b4278  /data/app/com.DefaultCompany.ArmNNObject-ZZiJpDLPTj7kfZWZo6ik8Q==/lib/arm64/libarmnn.so (armnn::Runtime::EnqueueWorkload(int, std::__ndk1::vector<std::__ndk1::pair<int, armnn::ConstTensor>, std::__ndk1::allocator<std::__ndk1::pair<int, armnn::ConstTensor> > > const&, std::__ndk1::vector<std::__ndk1::pair<int, armnn::Tensor>, std::__ndk1::allocator<std::__ndk1::pair<int, armnn::Tensor> > > const&)+48)
09-14 15:26:30.632 24359 24389 E AndroidRuntime:       #09 pc 0000000000003cac  /data/app/com.DefaultCompany.ArmNNObject-ZZiJpDLPTj7kfZWZo6ik8Q==/lib/arm64/libyoloDetector.so (detectObjects+200) (BuildId: e404cef578349f1e96d1437a5eb2458cec91605e)
```

I attached a .zip file with the code (please ignore random DERR macros used for debug). Model found [here](https://github.com/tsingjinyun/caffe-yolov2) under _Model files_ section (tried a number of them with the same result).

ArmNN libraries were compiled for _arm64-v8a_, using _IL2CPP_ scripting backend in Unity.

**SW stack:**
ARM NN: _v20.08_
ARM ComputeLibrary: _v20.08_
Android NDK: _r17b for Android 26 API standalone toolchain_
Boost: _v.1.64_
Protobuf: _v.3.5.2_
Caffe: _v1.0_
[armnn_caffe_infer.zip](https://github.com/ARM-software/armnn/files/5220624/armnn_caffe_infer.zip)

Any help much appreciated! Thanks!

**Comment by MikeJKelly** [comment]
Hi @the-promised-LAN 

What kind of device are you running this on? I don't have Unity but I managed to get a slightly modified version of your code compiling using just ArmNN and it worked fine (I used CpuRef instead of CpuAcc as I'm developing on an x86 device).

Best regards,
Mike

**Comment by MikeJKelly** [comment]
I was wondering what value of ```s_NetworkIdentifier``` are you're calling  
```armnn::Status ret = s_Runtime->EnqueueWorkload(s_NetworkIdentifier, inTensor, outTensor);``` with? 

NetworkId is an int so it should be easy to determine. If the network with that id isn't properly loaded then unordered_map will try to throw an out of range exception. In the stacktrace I see ```libc++_shared.__cxa_get_exception_ptr(__cxa_get_exception_ptr:0)``` so it's possible that your app may be running into some trouble when trying to throw that exception?

The first network loaded has a NetworkId of 0 so maybe initialize the variable with a different value to see if the network is being loaded and optimized correctly. Something like ```static armnn::NetworkId s_NetworkIdentifier = 42;``` would do.

**Comment by the-promised-LAN** [comment]
Hi @MikeJKelly,

Thanks for the feedback! I'm running this on a Samsung Galaxy S9 (Snapdragon 845, ARMv8.2, Cortex-A75/A55) and Android 10 (API 29). I'll try the different NetworkIDs and will reply back. Will CpuRef backend work on Cortex-A series, or GpuAcc on Adreno GPUs (I know it's Qualcomm, but just to confirm)?

Also when building Caffe, it's using x86_64 protobuf build, meanwhile ArmNN uses protobuf's arm64 build. Should I build Caffe with arm64 protobuf instead? 

Thanks again!

**Comment by MikeJKelly** [comment]
Hi @the-promised-LAN 

CpuRef will work on Cortex-A (it's our reference backend so it should work on just about anything but it's not optimized so the performance can be a lot worse than the CpuAcc). The GpuAcc will require an ARM GPU. 

Yes, you should build Caffe using the arm64 protobuf if you're deploying on an arm64. 

Best regards,
Mike

**Comment by the-promised-LAN** [comment]
Hi @MikeJKelly 

Tried setting different NetworkId and found the issue: EnqueueWorkload didn't see the loaded/optimized network. Even though network was loaded and optimized in one .cpp file (and saved in a global static variable), the same variable in other .cpp file calling EnqueueWorkload wasn't updated. Guessing there were some issues with linking the two files as the static variables were decoupled.

I put all the code in the same .cpp file and it worked perfectly. Thanks again for the help, closing this!
